rclcpp/services/async_client/main.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <cinttypes>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "example_interfaces/srv/add_two_ints.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+using AddTwoInts = example_interfaces::srv::AddTwoInts;
+using namespace std::chrono_literals;
+
+class ClientNode : public rclcpp::Node
+{
+public:
+  explicit ClientNode(const rclcpp::NodeOptions & options = rclcpp::NodeOptions{})
+  : Node("add_two_ints_async_client", options)
+  {
+    setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+    client_ = create_client<AddTwoInts>("add_two_ints");
+    // The client stores all pending requests internally, but those aren't cleaned up
+    // automatically if the server never responds.
+    // We create a timer that prunes all old requests each 5s, you can get the request ids
+    // of the requests that weren't completed here.
+    timer_ = this->create_wall_timer(
+      5s,
+      [this]() {
+        std::vector<int64_t> pruned_requests;
+        // Prune all requests older than 5s.
+        size_t n_pruned = this->client_->prune_requests_older_than(
+          std::chrono::system_clock::now() - 5s, &pruned_requests);
+        if (n_pruned) {
+          RCLCPP_INFO(
+            this->get_logger(),
+            "The server hasn't replied for more than 5s, %zu requests were discarded, "
+            "the discarded requests numbers are:",
+            n_pruned);
+          for (const auto & req_num : pruned_requests) {
+            RCLCPP_INFO(this->get_logger(), "\t%" PRId64, req_num);
+          }
+        }
+      });
+  }
+
+  bool
+  wait_for_service_server()
+  {
+    while (!client_->wait_for_service(std::chrono::seconds(1))) {
+      if (!rclcpp::ok()) {
+        RCLCPP_ERROR(this->get_logger(), "client interrupted while waiting for service to appear.");
+        return false;
+      }
+      RCLCPP_INFO(this->get_logger(), "waiting for service to appear...");
+    }
+    return true;
+  }
+
+  void queue_async_request(int64_t a, int64_t b)
```

Comment:
super nitpick: this isn't required but if you want to be consistent with the rest of the file it would be:

```suggestion
  void
  queue_async_request(int64_t a, int64_t b)
```